pkg/crds/crds.go

```diff
@@ -18,10 +18,185 @@ limitations under the License.
 package crds
 
 import (
+	"encoding/json"
+	"reflect"
+	"strings"
+
+	"github.com/ghodss/yaml"
 	"github.com/kubernetes-sigs/kustomize/pkg/loader"
+	"github.com/kubernetes-sigs/kustomize/pkg/transformers"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/kube-openapi/pkg/common"
 )
 
+type pathConfigs struct {
+	labelPathConfig          transformers.PathConfig
+	annotationPathConfig     transformers.PathConfig
+	prefixPathConfig         transformers.PathConfig
+	namereferencePathConfigs []transformers.ReferencePathConfig
+}
+
+func (p *pathConfigs) addLabelPathConfig(config transformers.PathConfig) {
+	if *(p.labelPathConfig.GroupVersionKind) == *config.GroupVersionKind {
+		p.labelPathConfig.Path = append(p.labelPathConfig.Path, config.Path...)
+	} else {
+		p.labelPathConfig = config
+	}
+}
+
+func (p *pathConfigs) addAnnotationPathConfig(config transformers.PathConfig) {
+	if *(p.annotationPathConfig.GroupVersionKind) == *config.GroupVersionKind {
+		p.annotationPathConfig.Path = append(p.labelPathConfig.Path, config.Path...)
+	} else {
+		p.annotationPathConfig = config
+	}
+}
+
+func (p *pathConfigs) addNamereferencePathConfig(config transformers.ReferencePathConfig) {
+	p.namereferencePathConfigs = transformers.MergeNameReferencePathConfigs(p.namereferencePathConfigs, config)
+}
+
+func (p *pathConfigs) addPrefixPathConfig(config transformers.PathConfig) {
+	if *(p.prefixPathConfig.GroupVersionKind) == *config.GroupVersionKind {
+		p.prefixPathConfig.Path = append(p.prefixPathConfig.Path, config.Path...)
+	} else {
+		p.prefixPathConfig = config
+	}
+}
+
 // RegisterCRDs parse CRD schemas from paths and update various pathConfigs
 func RegisterCRDs(loader loader.Loader, paths []string) error {
+	pathConfigs := []pathConfigs{}
+	for _, path := range paths {
+		pathConfig, err := registerCRD(loader, path)
+		if err != nil {
+			return err
+		}
+		pathConfigs = append(pathConfigs, pathConfig...)
+	}
+	addPathConfigs(pathConfigs)
+	return nil
+}
+
+// register CRD from one path
+func registerCRD(loader loader.Loader, path string) ([]pathConfigs, error) {
+	result := []pathConfigs{}
+
+	content, err := loader.Load(path)
+	if err != nil {
+		return result, err
+	}
+
+	var types map[string]common.OpenAPIDefinition
+	if content[0] == '{' {
+		err = json.Unmarshal(content, &types)
+	} else {
+		err = yaml.Unmarshal(content, &types)
+	}
+
+	crds := getCRDs(types)
+	for _, crd := range crds {
+		crdPathConfigs := pathConfigs{}
+		err = getCRDPathConfig(types, crd, crd, []string{}, &crdPathConfigs)
+		if err != nil {
+			return result, err
+		}
+		if !reflect.DeepEqual(crdPathConfigs, pathConfigs{}) {
+			result = append(result, crdPathConfigs)
+		}
+	}
+
+	return result, nil
+}
+
+// getCRDs get all CRD types
+func getCRDs(types map[string]common.OpenAPIDefinition) []string {
+	crds := []string{}
+
+	for typename, t := range types {
+		properties := t.Schema.SchemaProps.Properties
+		if _, ok := properties["kind"]; ok {
```

Comment:
IIRC "kind" in `properties` doesn't have the same meaning as in k8s.
We should think more carefully about how to handle GVK here.
Let's discuss this tomorrow morning.